### ISSUE TITLE
The "Collection was of a fixed size." exception when configuring options.

### ISSUE THREAD

**Issue by martintmk** [dotnet/runtime#77725]
### Description

Exception thrown from the collection when configuring and options. This bug happens when we configure the options using the `Configure` callback **AFTER** the `Bind` extension is called.

This happens when you are using non-concrete collection type in our options:

``` csharp
public class MyOptions
{
    public IList<string> MyList { get; set; } = new List<string>();
}
```

I suspect that the `Bind` extension materializes incorrect `IList` type that has a fixed size. and no more items can be added to it after that. 

We use this pattern across our codebase, and it breaks our scenarios completely. It blocks us from migrating to .NET 7. Full repro source [here](https://github.com/dotnet/runtime/files/9911897/ConsoleApp24.zip)

### Reproduction Steps

Run the following code in .NET 7.0 RC2 console app:

``` csharp
var config = new ConfigurationBuilder()
    .AddInMemoryCollection(new Dictionary<string, string?>() {{ "MyList:0", "item1" }})
    .Build();

var services = new ServiceCollection();
services
    .AddOptions<MyOptions>()
    .Bind(config)
    .Configure(o => o.MyList.Add("xxx"));

var val = services.BuildServiceProvider().GetRequiredService<IOptions<MyOptions>>().Value;

Console.WriteLine("List size: {0}", val.MyList.Count);

public class MyOptions
{
    public IList<string> MyList { get; set; } = new List<string>();
}
``` 

### Expected behavior

The code snippet above should work.

### Actual behavior

When trying to add a new item to a list the list throws the following exception:

``` text
Unhandled exception. System.NotSupportedException: Collection was of a fixed size.
```

### Known Workarounds

Change the collection to a concrete type. i.e. `IList<string>` -> `List<string>` 

### Configuration

Running on .NET 7 using the default console template.

**Comment by geeknoid** [comment]
Very important thing to note. This is new behavior in .NET 7, this bug was not present in previous versions of .NET. This change in .NET behavior breaks a lot of our code and our customer's code.

**Comment by tarekgh** [comment]
We'll take a look soon at this one. Thanks for your report.

**Comment by tarekgh** [comment]
CC @SteveDunn as FYI.

**Comment by SteveDunn** [comment]
> CC @SteveDunn as FYI.

Hi, happy to fix, but unfortunately I'm not free until the weekend. Will this go into .NET 7?

**Comment by stephentoub** [comment]
What change regressed this, @tarekgh / @SteveDunn?

**Comment by tarekgh** [comment]
@stephentoub we are going to investigate as I mentioned in https://github.com/dotnet/runtime/issues/77725#issuecomment-1299024482

**Comment by tarekgh** [comment]
https://github.com/dotnet/runtime/pull/73267

**Comment by eerhardt** [comment]
Before https://github.com/dotnet/runtime/pull/68133, the checks were fairly simple:

https://github.com/dotnet/runtime/blob/fe217956fd128732279c1d01d854d025f8e225d1/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L380-L396

This would fall to `BindCollection`, which would just add to the existing object instance.

Now, we are checking upfront if the property's type is compatible with an array, and if it is, copying the current instance's values + the config values into a new array:

https://github.com/dotnet/runtime/blob/78cf32aff96ab5b10aef662a4c6889e7914154f7/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L302-L307

This is incorrect for "mutable" collection types (like `IList<T>` and `ICollection<T>`). For those types, we should just be appending to the existing instance, if there is one.

**Comment by tarekgh** [comment]
@eerhardt I am working on a fix. We need to first create instance `using List<T>` before binding.  anyway, let you know what I'll get.

**Comment by tarekgh** [comment]
I opened https://github.com/dotnet/runtime/pull/77857 to fix this issue.

**Comment by geeknoid** [comment]
I presume the fix isn't going to make .NET 7's initial release?

**Comment by stephentoub** [comment]
Correct. It would be in a subsequent servicing release.

**Comment by tarekgh** [comment]
This is fixed by the PR https://github.com/dotnet/runtime/pull/78118. The fix will be in the next 7.0 service release.
